File: core/src/main/scala/com/graphbrain/eco/nodes/NlpFun.scala
package com.graphbrain.eco.nodes

import com.graphbrain.eco._
import scala.Boolean
import com.graphbrain.eco.nodes.NlpFunType.NlpFunType
import com.graphbrain.eco.NodeType.NodeType
import com.graphbrain.eco.NodeType

object NlpFunType extends Enumeration {
  type NlpFunType = Value
  val IS_POS,
    IS_POSPRE,
    ARE_POS,
    ARE_POSPRE,
    CONTAINS_POS,
    CONTAINS_POSPRE,
    IS_LEMMA = Value
}

class NlpFun(val funType: NlpFunType, params: Array[ProgNode], lastTokenPos: Int= -1) extends FunNode(params, lastTokenPos) {

  override val label = funType match {
    case NlpFunType.IS_POS => "is-pos"
    case NlpFunType.IS_POSPRE => "is-pos-pre"
    case NlpFunType.ARE_POS => "are-pos"
    case NlpFunType.ARE_POSPRE => "are-pos-pre"
    case NlpFunType.CONTAINS_POS => "contains-pos"
    case NlpFunType.CONTAINS_POSPRE => "contains-pos-pre"
    case NlpFunType.IS_LEMMA => "is-lemma"
  }

  override def ntype: NodeType = NodeType.Boolean

  private def matchPoses(word: Word, pre: Boolean, poses: Array[String]): Boolean = {
    for (pos <- poses)
      if (pre)
        if (word.pos.startsWith(pos)) return true
      else
        if (word.pos == pos) return true

    false
  }

  private def matchPoses(words: Words, pre: Boolean, ctxt: Context): Boolean = {
    val poses = params.drop(1).map(ctxt.getRetString)

    for (word <- words.words)
      if (!matchPoses(word, pre, poses)) return false

    true
  }

  private def containsPoses(words: Words, pre: Boolean, ctxt: Context): Boolean = {
    val poses = params.drop(1).map(ctxt.getRetString)

    for (word <- words.words)
      if (matchPoses(word, pre, poses)) return true

    false
  }

  override def booleanValue(ctxts: Contexts) = {
    params(0).wordsValue(ctxts)
    params.drop(1).map(_.stringValue(ctxts))

    for (c <- ctxts.ctxts) {
      val words = c.getRetWords(params(0))

      funType match {
        case NlpFunType.IS_POS => {
          c.setRetBoolean(this, if (words.length != 1)
            false
          else
            matchPoses(words, pre = false, c))
        }
        case NlpFunType.IS_POSPRE => {
          c.setRetBoolean(this, if (words.length != 1)
            false
          else
            matchPoses(words, pre = true, c))
        }
        case NlpFunType.ARE_POS => {
          c.setRetBoolean(this, if (words.length == 0)
            false
          else
            matchPoses(words, pre = false, c))
        }
        case NlpFunType.ARE_POSPRE => {
          c.setRetBoolean(this, if (words.length == 0)
            false
          else
            matchPoses(words, pre = true, c))
        }
        case NlpFunType.CONTAINS_POS => {
          c.setRetBoolean(this, if (words.length == 0)
            false
          else
            containsPoses(words, pre = false, c))
        }
        case NlpFunType.CONTAINS_POSPRE => {
          c.setRetBoolean(this, if (words.length == 0)
            false
          else
            containsPoses(words, pre = true, c))
        }
        case NlpFunType.IS_LEMMA => {
          c.setRetBoolean(this, if (words.length != 1)
            false
          else
            words.words(0).lemma == c.getRetString(params(1)))
        }
      }
    }
  }
}

File: core/src/main/java/com/graphbrain/eco/nodes/NlpFun.java
package com.graphbrain.eco.nodes;

import com.graphbrain.eco.*;
import com.graphbrain.eco.NodeType;

public class NlpFun extends FunNode {

    public enum NlpFunType {
        IS_POS, IS_POSPRE, ARE_POS, ARE_POSPRE, CONTAINS_POS, CONTAINS_POSPRE, IS_LEMMA
    }

    private NlpFunType fun;

    public NlpFun(NlpFunType fun, ProgNode[] params, int lastTokenPos) {
        super(params, lastTokenPos);
        this.fun = fun;
    }

    public NlpFun(NlpFunType fun, ProgNode[] params) {
        this(fun, params, -1);
    }

    @Override
    public String label() {
        switch(fun) {
            case IS_POS: return "is-pos";
            case IS_POSPRE: return "is-pos-pre";
            case ARE_POS: return "are-pos";
            case ARE_POSPRE: return "are-pos-pre";
            case CONTAINS_POS: return "contains-pos";
            case CONTAINS_POSPRE: return "contains-pos-pre";
            case IS_LEMMA: return "is-lemma";
        }
        // error
        return "?";
    }

    @Override
    public NodeType ntype(){return NodeType.Boolean;}

    private boolean matchPoses(Word word, boolean pre, String[] poses) {
        for (String pos : poses)
            if (pre)
                if (word.getPos().startsWith(pos))
                    return true;
            else
                if (word.getPos().equals(pos))
                    return true;

        return false;
    }

    private boolean matchPoses(Words words, boolean pre, Context ctxt) {
        String[] poses = new String[params.length - 1];
        for (int i = 1; i < params.length; i++) {
            poses[i - 1] = ctxt.getRetString(params[i]);
        }

        for (Word word : words.getWords())
            if (!matchPoses(word, pre, poses))
                return false;

        return true;
    }

    private boolean containsPoses(Words words, boolean pre, Context ctxt) {
        String[] poses = new String[params.length - 1];
        for (int i = 1; i < params.length; i++) {
            poses[i - 1] = ctxt.getRetString(params[i]);
        }

        for (Word word : words.getWords())
            if (matchPoses(word, pre, poses))
                return true;

        return false;
    }

    @Override
    public void booleanValue(Contexts ctxts) {
        params[0].wordsValue(ctxts);

        for (int i = 1; i < params.length; i++) {
            params[i].stringValue(ctxts);
        }

        for (Context c : ctxts.getCtxts()) {
            Words words = c.getRetWords(params[0]);

            boolean ret;
            switch(fun) {
                case IS_POS:
                    ret = words.length() == 1 && matchPoses(words, false, c);
                    c.setRetBoolean(this, ret);
                    break;
                case IS_POSPRE:
                    ret = words.length() == 1 && matchPoses(words, true, c);
                    c.setRetBoolean(this, ret);
                    break;
                case ARE_POS:
                    ret = words.length() != 0 && matchPoses(words, false, c);
                    c.setRetBoolean(this, ret);
                    break;
                case ARE_POSPRE:
                    ret = words.length() != 0 && matchPoses(words, true, c);
                    c.setRetBoolean(this, ret);
                    break;
                case CONTAINS_POS:
                    ret = words.length() != 0 && containsPoses(words, false, c);
                    c.setRetBoolean(this, ret);
                    break;
                case CONTAINS_POSPRE:
                    ret = words.length() != 0 && containsPoses(words, true, c);
                    c.setRetBoolean(this, ret);
                    break;
                case IS_LEMMA:
                    ret = words.length() == 1 && words.getWords()[0].getLemma().equals(c.getRetString(params[1]));
                    c.setRetBoolean(this, ret);
                    break;
            }
        }
    }
}